processing_provider/datagen.py

```diff
@@ -0,0 +1,184 @@
+        should be localised.
+        """
+        return self.tr('QDeepLandia')
+
+    def groupId(self):
+        """
+        Returns the unique ID of the group this algorithm belongs to. This
+        string should be fixed for the algorithm, and must not be localised.
+        The group id should be unique within each provider. Group id should
+        contain lowercase alphanumeric characters only and no spaces or other
+        formatting characters.
+        """
+        return 'QDeepLandia'
+
+    def shortHelpString(self):
+        """
+        Returns a localised short helper string for the algorithm. This string
+        should provide a basic description about what the algorithm does and the
+        parameters and outputs associated with it..
+        """
+        return self.tr("Preprocess layer into predictable tiles")
+
+    def initAlgorithm(self, config=None):
+        """
+        Here we define the inputs and output of the algorithm, along
+        with some other properties.
+        """
+
+        # We add the input vector features source. It can have any kind of
+        # geometry.
+        self.addParameter(
+            QgsProcessingParameterRasterLayer(
+                self.INPUT,
+                self.tr('Input layer')
+            )
+        )
+
+        # We add a feature sink in which to store our processed features (this
+        # usually takes the form of a newly created vector layer when the
+        # algorithm is run in QGIS).
+        self.addParameter(
+            QgsProcessingParameterString(
+                self.DATASET,
+                self.tr('Dataset name')
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterNumber(
+                self.SHAPE,
+                self.tr('Number of pixel for the side of tiles'),
+                type = QgsProcessingParameterNumber.Integer,
+                defaultValue = 512,
+                minValue = 16
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFolderDestination(
+                self.OUTPUT,
+                self.tr('Output folder')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+        """
+        Here is where the processing itself takes place.
+        """
+
+        raster_in = self.parameterAsRasterLayer(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        dest_path = self.parameterAsString(
+            parameters,
+            self.OUTPUT,
+            context
+        )
+
+        dataset = self.parameterAsString(
+            parameters,
+            self.DATASET,
+            context
+        )
+
+        shape = self.parameterAsInt(
+            parameters,
+            self.SHAPE,
+            context
+        )
+
+        path=''
+        for i in [dest_path, dataset, 'input','testing','images']:
+            path = os.path.join(path, i)
```

Comment:
`os.path` does not seem very practical in this use case... Maybe we should consider `pathlib.Path` (Python3) in the future.